File: sdks/node/src/proxy.ts
/**
 * PolicyProxy — wraps an LLM client to route tool calls through TrueBearing.
 *
 * This module owns the lifecycle of the TrueBearing proxy subprocess (when running
 * in subprocess mode) and the header injection that associates every outbound
 * request with a session ID and a signed agent JWT.
 *
 * Callers interact primarily with PolicyProxy.create(). resolveJwt, findFreePort,
 * startSubprocess, and waitForReady are exported for testability.
 */

import * as childProcess from "child_process";
import * as fs from "fs";
import * as http from "http";
import * as net from "net";
import * as os from "os";
import * as path from "path";
import { randomUUID } from "crypto";

/** Options for constructing a PolicyProxy. */
export interface PolicyProxyOptions {
  /**
   * Path to a TrueBearing policy YAML file. Required in subprocess mode
   * (when proxyUrl is not provided).
   */
  policy?: string;
  /**
   * URL of a running TrueBearing proxy. When given, no subprocess is spawned.
   * This is the sidecar deployment model.
   */
  proxyUrl?: string;
  /**
   * The agent JWT string. Takes precedence over the TRUEBEARING_AGENT_JWT
   * environment variable and the key file.
   */
  agentJwt?: string;
  /**
   * Agent name whose JWT is read from ~/.truebearing/keys/<agentName>.jwt.
   * Used only when agentJwt is not provided and the env var is absent.
   */
  agentName?: string;
  /**
   * Explicit session ID to continue a prior session. Defaults to a fresh
   * UUID v4 when not provided.
   */
  sessionId?: string;
  /**
   * Upstream MCP server URL passed to truebearing serve in subprocess mode.
   * Defaults to http://localhost:8080.
   */
  upstream?: string;
}

/** Maximum milliseconds to wait for the proxy subprocess to become ready. */
const READINESS_TIMEOUT_MS = 30_000;
/** Milliseconds between /health polls while waiting for subprocess readiness. */
const POLL_INTERVAL_MS = 250;

/**
 * Wraps an LLM client to route all MCP tool calls through the TrueBearing proxy.
 *
 * Two operating modes:
 *
 * **Subprocess mode** (use PolicyProxy.create() with a policy path): spawns
 * `truebearing serve` on a random local port, waits for the /health endpoint to
 * return 200, then reconfigures the wrapped client to use the proxy URL. The
 * subprocess is terminated when close() is called.
 *
 * **Explicit proxy mode** (pass proxyUrl): no subprocess is spawned. The proxy
 * must already be running (e.g. as a sidecar container).
 *
 * Header injection: every outbound request carries:
 *   - `Authorization: Bearer <jwt>` — the agent JWT issued by
 *     `truebearing agent register`.
 *   - `X-TrueBearing-Session-ID: <uuid4>` — the session identifier that
 *     scopes sequence evaluation and budget tracking to this agent run.
 *
 * @example
 * // Subprocess mode:
 * import { PolicyProxy } from '@mercator/truebearing';
 * import Anthropic from '@anthropic-ai/sdk';
 *
 * const proxy = await PolicyProxy.create(new Anthropic(), { policy: './policy.yaml' });
 * try {
 *   const response = await (proxy.client as Anthropic).messages.create({ ... });
 * } finally {
 *   proxy.close();
 * }
 *
 * @example
 * // Explicit proxy mode — proxy already running as a sidecar:
 * const proxy = await PolicyProxy.create(new Anthropic(), {
 *   proxyUrl: 'http://localhost:7773',
 *   agentJwt: process.env.TRUEBEARING_AGENT_JWT,
 * });
 */
export class PolicyProxy {
  /** @internal exposed for testing */
  readonly _sessionId: string;
  /** @internal exposed for testing */
  _proc: childProcess.ChildProcess | null;
  /** @internal exposed for testing */
  readonly _jwt: string | null;
  /** @internal exposed for testing */
  readonly _proxyUrl: string;
  /** @internal exposed for testing */
  _client: unknown;

  private constructor(
    client: unknown,
    proxyUrl: string,
    jwt: string | null,
    sessionId: string,
    proc: childProcess.ChildProcess | null,
  ) {
    this._sessionId = sessionId;
    this._jwt = jwt;
    this._proxyUrl = proxyUrl;
    this._proc = proc;
    this._client = configureClient(client, proxyUrl, jwt, sessionId);
  }

  /**
   * Create a PolicyProxy, optionally spawning a truebearing serve subprocess.
   *
   * In subprocess mode (policy provided, no proxyUrl), this method:
   * 1. Finds a free local port.
   * 2. Spawns `truebearing serve` on that port.
   * 3. Polls /health until the proxy is ready.
   * 4. Returns a PolicyProxy whose .client is reconfigured to route through the proxy.
   *
   * In explicit proxy mode (proxyUrl provided), no subprocess is spawned and
   * no async work is required. The method still returns a Promise for API
   * consistency.
   *
   * Throws if policy is absent in subprocess mode, or if the proxy does not
   * become ready within 30 seconds.
   */
  static async create(
    client: unknown,
    options: PolicyProxyOptions = {},
  ): Promise<PolicyProxy> {
    const { policy, proxyUrl, agentJwt, agentName, sessionId, upstream } =
      options;
    const resolvedSessionId = sessionId ?? randomUUID();
    const jwt = resolveJwt(agentJwt, agentName);

    if (proxyUrl !== undefined) {
      return new PolicyProxy(
        client,
        proxyUrl.replace(/\/$/, ""),
        jwt,
        resolvedSessionId,
        null,
      );
    }

    if (policy === undefined) {
      throw new Error(
        "'policy' is required when not providing an explicit 'proxyUrl'. " +
          "Either pass { policy: './policy.yaml' } (subprocess mode) or " +
          "{ proxyUrl: 'http://localhost:7773' } (explicit proxy mode).",
      );
    }

    const port = await findFreePort();
    const resolvedProxyUrl = `http://localhost:${port}`;
    const proc = startSubprocess(port, policy, upstream);
    const instance = new PolicyProxy(
      client,
      resolvedProxyUrl,
      jwt,
      resolvedSessionId,
      proc,
    );

    try {
      await instance._waitForReady();
    } catch (err) {
      proc.kill();
      throw err;
    }

    return instance;
  }

  /**
   * Poll /health until the proxy responds 200 or the timeout expires.
   *
   * Extracted as an instance method so tests can spy on it without needing to
   * intercept raw HTTP calls.
   *
   * @internal
   */
  async _waitForReady(): Promise<void> {
    const healthUrl = `${this._proxyUrl}/health`;
    const deadline = Date.now() + READINESS_TIMEOUT_MS;

    while (Date.now() < deadline) {
      const ok = await checkHealth(healthUrl);
      if (ok) return;
      await sleep(POLL_INTERVAL_MS);
    }

    throw new Error(
      `TrueBearing proxy did not become ready within ${READINESS_TIMEOUT_MS / 1000}s. ` +
        "Verify that the 'truebearing' binary is on PATH, the policy file is valid, " +
        `and no other process is occupying the port (${this._proxyUrl}).`,
    );
  }

  /** The reconfigured wrapped client, ready to route requests through the proxy. */
  get client(): unknown {
    return this._client;
  }

  /** Terminate the proxy subprocess if one was spawned by this instance. */
  close(): void {
    if (this._proc !== null) {
      this._proc.kill("SIGTERM");
      this._proc = null;
    }
  }

  /**
   * Support `await using proxy = await PolicyProxy.create(...)` via TC39
   * explicit resource management (TypeScript 5.2+).
   */
  [Symbol.asyncDispose](): Promise<void> {
    this.close();
    return Promise.resolve();
  }
}

// ---------------------------------------------------------------------------
// Module-level helpers (exported for testability)
// ---------------------------------------------------------------------------

/**
 * Return the agent JWT from the first available source.
 *
 * Priority:
 * 1. agentJwt argument (explicit).
 * 2. TRUEBEARING_AGENT_JWT environment variable.
 * 3. ~/.truebearing/keys/<agentName>.jwt file (when agentName is given).
 */
export function resolveJwt(
  agentJwt: string | undefined,
  agentName: string | undefined,
): string | null {
  if (agentJwt) return agentJwt;
  const envVal = process.env["TRUEBEARING_AGENT_JWT"];
  if (envVal) return envVal;
  if (agentName) {
    const keyPath = path.join(
      os.homedir(),
      ".truebearing",
      "keys",
      `${agentName}.jwt`,
    );
    return fs.readFileSync(keyPath, "utf8").trim();
  }
  return null;
}

/**
 * Return an available TCP port on localhost.
 *
 * The server is closed immediately after the OS assigns a port. There is a
 * small TOCTOU window before the subprocess binds the port, but it is
 * acceptable for local development use.
 */
export function findFreePort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const srv = net.createServer();
    srv.on("error", reject);
    srv.listen(0, "127.0.0.1", () => {
      const port = (srv.address() as net.AddressInfo).port;
      srv.close((err) => {
        if (err) reject(err);
        else resolve(port);
      });
    });
  });
}

/**
 * Spawn `truebearing serve` on the given port and return the process handle.
 *
 * Output is suppressed because operators observe proxy behaviour via
 * `truebearing audit query`, not subprocess stdout/stderr.
 */
export function startSubprocess(
  port: number,
  policy: string,
  upstream: string | undefined,
): childProcess.ChildProcess {
  const upstreamUrl = upstream ?? "http://localhost:8080";
  return childProcess.spawn(
    "truebearing",
    [
      "serve",
      "--port",
      String(port),
      "--policy",
      policy,
      "--upstream",
      upstreamUrl,
    ],
    { stdio: "ignore" },
  );
}

/** Issue a GET to url and return true iff the response status is 200. */
function checkHealth(url: string): Promise<boolean> {
  return new Promise((resolve) => {
    const req = http.get(url, { timeout: 2000 }, (res) => {
      resolve(res.statusCode === 200);
      // Consume the response body to free the socket.
      res.resume();
    });
    req.on("error", () => resolve(false));
    req.on("timeout", () => {
      req.destroy();
      resolve(false);
    });
  });
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Reconfigure client to route through the proxy and inject required headers.
 *
 * Returns a new client instance for known SDK types, or the original client
 * unchanged when the SDK is unrecognised.
 *
 * Design: SDK detection uses duck typing on the withOptions method so the SDK
 * packages remain optional runtime dependencies. PolicyProxy works without them
 * when proxyUrl is supplied explicitly and the caller configures headers
 * themselves. This mirrors the Python SDK's try/import pattern.
 */
function configureClient(
  client: unknown,
  proxyUrl: string,
  jwt: string | null,
  sessionId: string,
): unknown {
  const extraHeaders: Record<string, string> = {};
  if (jwt) {
    extraHeaders["Authorization"] = `Bearer ${jwt}`;
  }
  extraHeaders["X-TrueBearing-Session-ID"] = sessionId;

  // Anthropic SDK (and any compatible SDK) exposes withOptions() to return a
  // new client instance with overridden options. Use duck typing so the package
  // stays free of runtime dependencies.
  if (
    client !== null &&
    typeof client === "object" &&
    "withOptions" in client &&
    typeof (client as { withOptions: unknown }).withOptions === "function"
  ) {
    return (
      client as {
        withOptions: (opts: {
          baseURL: string;
          defaultHeaders: Record<string, string>;
        }) => unknown;
      }
    ).withOptions({ baseURL: proxyUrl, defaultHeaders: extraHeaders });
  }

  // Unrecognised client: return unchanged. The caller is responsible for
  // header injection in this case (e.g. via a custom fetch wrapper).
  return client;
}

File: sdks/python/src/truebearing/_proxy.py
"""PolicyProxy — wraps an LLM client to route tool calls through TrueBearing.

This module owns the lifecycle of the TrueBearing proxy subprocess (when running
in subprocess mode) and the header injection that associates every outbound
request with a session ID and a signed agent JWT.

Callers interact only with PolicyProxy. _find_free_port is module-level so it
can be patched in tests without needing to subclass PolicyProxy.
"""

import os
import socket
import subprocess
import time
import urllib.error
import urllib.request
import uuid
from pathlib import Path
from typing import Optional


class PolicyProxy:
    """Wraps an LLM client to route all MCP tool calls through the TrueBearing proxy.

    Two operating modes:

    Subprocess mode (default): PolicyProxy spawns ``truebearing serve`` on a
    random local port, waits for the /health endpoint to return 200, then
    reconfigures the wrapped client to use the proxy URL. The subprocess is
    terminated when the PolicyProxy is closed via context manager exit or
    garbage collection.

    Explicit proxy mode: caller provides ``proxy_url`` pointing to a proxy
    that is already running. No subprocess is spawned. This is the sidecar
    deployment model.

    Header injection: every outbound request carries:
      - ``Authorization: Bearer <jwt>`` — the agent JWT issued by
        ``truebearing agent register``.
      - ``X-TrueBearing-Session-ID: <uuid4>`` — the session identifier that
        scopes sequence evaluation and budget tracking to this agent run. One
        PolicyProxy instance equals one TrueBearing session.

    Usage (subprocess mode)::

        from truebearing import PolicyProxy
        import anthropic

        client = PolicyProxy(anthropic.Anthropic(), policy="./policy.yaml")
        # client.messages.create(...) routes through the proxy.

    Usage (explicit proxy URL)::

        client = PolicyProxy(
            anthropic.Anthropic(),
            proxy_url="http://localhost:7773",
            agent_jwt=os.environ["TRUEBEARING_AGENT_JWT"],
        )
    """

    # Maximum seconds to wait for the proxy subprocess to become ready.
    _READINESS_TIMEOUT_S = 30
    # Seconds between /health polls while waiting for subprocess readiness.
    _POLL_INTERVAL_S = 0.25

    def __init__(
        self,
        client,
        *,
        policy: Optional[str] = None,
        proxy_url: Optional[str] = None,
        agent_jwt: Optional[str] = None,
        agent_name: Optional[str] = None,
        session_id: Optional[str] = None,
        upstream: Optional[str] = None,
    ) -> None:
        """Initialise the PolicyProxy.

        Args:
            client: The LLM client to wrap (e.g. ``anthropic.Anthropic()``).
            policy: Path to a TrueBearing policy YAML file. Required when
                ``proxy_url`` is not provided (subprocess mode).
            proxy_url: URL of a running TrueBearing proxy. When given, no
                subprocess is spawned.
            agent_jwt: The agent JWT string. Takes precedence over
                ``TRUEBEARING_AGENT_JWT`` env var and the key file.
            agent_name: Agent name whose JWT is read from
                ``~/.truebearing/keys/<agent_name>.jwt``. Used only when
                ``agent_jwt`` is not provided and the env var is absent.
            session_id: Explicit session ID to continue a prior session.
                Defaults to a fresh ``uuid4`` when not provided.
            upstream: Upstream MCP server URL passed to ``truebearing serve``
                in subprocess mode. Defaults to ``http://localhost:8080``.
        """
        self._session_id: str = session_id or str(uuid.uuid4())
        # _proc is set only in subprocess mode; None means explicit proxy URL.
        self._proc: Optional[subprocess.Popen] = None

        self._jwt: Optional[str] = _resolve_jwt(agent_jwt, agent_name)

        if proxy_url is not None:
            self._proxy_url = proxy_url.rstrip("/")
        else:
            if policy is None:
                raise ValueError(
                    "'policy' is required when not providing an explicit 'proxy_url'. "
                    "Either pass policy='./policy.yaml' (subprocess mode) or "
                    "proxy_url='http://localhost:7773' (explicit proxy mode)."
                )
            port = _find_free_port()
            self._proxy_url = f"http://localhost:{port}"
            self._proc = _start_subprocess(port, policy, upstream)
            self._wait_for_ready()

        # Reconfigure the wrapped client to use the proxy URL and inject headers.
        self._client = _configure_client(client, self._proxy_url, self._jwt, self._session_id)

    def _wait_for_ready(self) -> None:
        """Poll /health until the proxy responds 200 or the timeout expires.

        Terminates the subprocess and raises RuntimeError on timeout so callers
        receive a clear error rather than a silent hang.
        """
        health_url = f"{self._proxy_url}/health"
        deadline = time.monotonic() + self._READINESS_TIMEOUT_S
        while time.monotonic() < deadline:
            try:
                with urllib.request.urlopen(health_url, timeout=2) as resp:
                    if resp.status == 200:
                        return
            except (urllib.error.URLError, OSError):
                # Proxy not yet listening; keep polling.
                pass
            time.sleep(self._POLL_INTERVAL_S)

        # Timed out — kill the orphan subprocess before raising.
        if self._proc is not None:
            self._proc.terminate()
            self._proc = None
        raise RuntimeError(
            f"TrueBearing proxy did not become ready within {self._READINESS_TIMEOUT_S}s. "
            "Verify that the 'truebearing' binary is on PATH, the policy file is valid, "
            f"and no other process is occupying the port ({self._proxy_url})."
        )

    def __enter__(self) -> "PolicyProxy":
        return self

    def __exit__(self, *args: object) -> None:
        self._shutdown()

    def __del__(self) -> None:
        # Best-effort cleanup: do not raise from __del__.
        try:
            self._shutdown()
        except Exception:  # noqa: BLE001
            pass

    def _shutdown(self) -> None:
        """Terminate the proxy subprocess if one was spawned by this instance."""
        if self._proc is not None:
            self._proc.terminate()
            try:
                self._proc.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self._proc.kill()
                self._proc.wait()
            self._proc = None

    def __getattr__(self, name: str) -> object:
        """Delegate all attribute lookups to the reconfigured wrapped client.

        This makes PolicyProxy transparent: ``proxy.messages.create(...)``
        behaves identically to ``client.messages.create(...)`` but routes
        through the TrueBearing proxy with injected headers.
        """
        return getattr(self._client, name)


# ---------------------------------------------------------------------------
# Module-level helpers (separate functions so they are patchable in tests)
# ---------------------------------------------------------------------------


def _resolve_jwt(agent_jwt: Optional[str], agent_name: Optional[str]) -> Optional[str]:
    """Return the agent JWT from the first available source.

    Priority:
    1. ``agent_jwt`` kwarg (explicit).
    2. ``TRUEBEARING_AGENT_JWT`` environment variable.
    3. ``~/.truebearing/keys/<agent_name>.jwt`` file (when ``agent_name`` is given).
    """
    if agent_jwt:
        return agent_jwt
    env_val = os.environ.get("TRUEBEARING_AGENT_JWT")
    if env_val:
        return env_val
    if agent_name:
        key_path = Path.home() / ".truebearing" / "keys" / f"{agent_name}.jwt"
        return key_path.read_text().strip()
    return None


def _find_free_port() -> int:
    """Return an available TCP port on localhost.

    The socket is closed immediately after binding so the port is free for
    the subprocess to bind. There is a small TOCTOU window, but it is
    acceptable for local development use.
    """
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind(("", 0))
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        return s.getsockname()[1]


def _start_subprocess(port: int, policy: str, upstream: Optional[str]) -> "subprocess.Popen[bytes]":
    """Spawn ``truebearing serve`` and return the process handle.

    Output is suppressed because operators observe proxy behaviour via
    ``truebearing audit query``, not subprocess stdout/stderr. Keeping output
    suppressed also prevents the subprocess from interfering with the host
    process's terminal.
    """
    upstream_url = upstream or "http://localhost:8080"
    cmd = [
        "truebearing",
        "serve",
        "--port",
        str(port),
        "--policy",
        policy,
        "--upstream",
        upstream_url,
    ]
    return subprocess.Popen(
        cmd,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )


def _configure_client(
    client: object,
    proxy_url: str,
    jwt: Optional[str],
    session_id: str,
) -> object:
    """Reconfigure ``client`` to route through the proxy and inject required headers.

    Supported client types:
      - ``anthropic.Anthropic`` (sync)
      - ``anthropic.AsyncAnthropic`` (async)

    Passing any other type raises ``ValueError`` with actionable instructions
    pointing the caller to the manual ``base_url`` workaround and the
    integrations documentation.

    Design: silent pass-through for unrecognised SDKs was deliberately removed.
    An unrecognised client returned unchanged would route tool calls directly to
    the upstream, bypassing TrueBearing enforcement entirely, with no indication
    of failure. A loud ValueError at construction time is the correct failure
    mode for a security proxy (fail-closed).
    """
    extra_headers: dict[str, str] = {}
    if jwt:
        extra_headers["Authorization"] = f"Bearer {jwt}"
    extra_headers["X-TrueBearing-Session-ID"] = session_id

    # Anthropic SDK (anthropic>=0.40): with_options returns a new client
    # instance with the supplied overrides applied. base_url routes all API
    # calls through the TrueBearing proxy.
    try:
        import anthropic  # type: ignore[import]

        if isinstance(client, anthropic.Anthropic):
            return client.with_options(
                base_url=proxy_url,
                default_headers=extra_headers,
            )
        if isinstance(client, anthropic.AsyncAnthropic):
            return client.with_options(
                base_url=proxy_url,
                default_headers=extra_headers,
            )
    except ImportError:
        pass

    raise ValueError(
        f"Unsupported client type: {type(client).__name__!r}. "
        "TrueBearing currently supports anthropic.Anthropic and anthropic.AsyncAnthropic. "
        "To use a different SDK, configure the proxy URL on your client manually: "
        f"client = YourClient(base_url='{proxy_url}'). "
        "See https://docs.mercator.dev/integrations for the full list of supported clients."
    )
